### DISCUSSION TITLE
<Scripts /> from @remix-run/react package are causing Hydration errors

### DISCUSSION THREAD

**Discussion by Blaumaus** [remix-run/remix#6668] [Q&A]
Hey! The title is self-explanatory.

root.tsx:
```
...
      <body
        className={clsx({
          'bg-white': theme === 'light',
          'bg-slate-900': theme === 'dark',
        })}
      >
        <Provider store={store}>
          <AlertProvider template={AlertTemplate} {...options}>
            <AppWrapper
              ssrTheme={theme}
              ssrAuthenticated={isAuthed}
            />
          </AlertProvider>
        </Provider>
        <ClientOnly>
          {() => (
            <ScrollRestoration />
          )}
        </ClientOnly>
        <Scripts />
        <LiveReload />
      </body>
...
```

The Scripts tag injects dozens of elements like `<link rel="modulepreload" href="/build/_shared/chunk-EMM7DE2W.js">` into the final HTML page.
This causes the following error:
```
Warning: Expected server HTML to contain a matching <link> in <body>.
link
Scripts@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:4856:7
body
html
App2@http://localhost:3000/build/root-GSFC7ITE.js:2235:7
RemixRoute@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:4389:20
RenderedRoute@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:2853:7
RenderErrorBoundary@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3244:9
DataRoutes@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3091:7
Router@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3105:7
RouterProvider@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3041:7
RemixErrorBoundary@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3998:5
RemixBrowser@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:5515:40
I18nextProvider@http://localhost:3000/build/_shared/chunk-UE4NS74C.js:283:14
```

Have any of you encountered this error? How to fix it?

Thanks.

**Comment by xHomu** [comment]
`<Scripts />` is how Remix loads the client side js bundles, enable client-side navigation, and attach event handlers. Scripts modulepreload won't trigger hydration mismatch, what would is if you have some client state that would result in a different JSX return in the initial render. Are you sure there's no, say, a theme mismatch between server and client?

Remember, Remix works by rendering everything on the Server first, and hydration fails when Client's initial render differs from the HTML that the server passed down. Anything that could change the initial render between Client and Server should probably be wrapped in ClientOnly.


When all else fails I open devtools and copy the HTML body to diffchecker.com, then disable javascript (Ctrl+Shift+P) and refresh, copy the body again and see if you can spot any difference. 

You can safely ignore the stuff near the bottom relating to link modulepreload and remixContext. Remember to re-enable Javascript afterwards!

**Reply by Blaumaus** [reply]
Here's the difference between SSR and CSR of a simple page I'm trying to render: https://www.diffchecker.com/ha5Elvo9/, my code is the same for both client and server side.
But the parts Remix is supposed to handle are different:
1. Remix adds comments into SSR HTML (this is probably not causing hydration errors but anyway)
2. Added an empty `media=""` to Tailwind import for some reason
3. `<link rel="modulepreload"....`

And here's an error I'm getting:
```
Warning: Expected server HTML to contain a matching <script> in <body>.
script
ScrollRestoration2@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:5548:28
body
html
App2@http://localhost:3000/build/root-5DDHFXQ7.js:1694:7
RemixRoute@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:4389:20
RenderedRoute@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:2853:7
RenderErrorBoundary@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3244:9
DataRoutes@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3091:7
Router@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3105:7
RouterProvider@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3041:7
RemixErrorBoundary@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:3998:5
RemixBrowser@http://localhost:3000/build/_shared/chunk-KHLNOZZL.js:5515:40
I18nextProvider@http://localhost:3000/build/_shared/chunk-UE4NS74C.js:283:14
```

I literally have no idea what's causing that issue.

**Reply by xHomu** [reply]
If you don't mind, I'll take a look at https://github.com/Swetrix/swetrix-fe tomorrow and see what I find.

**Comment by xHomu** [comment]
@Blaumaus 

I went through your front page and identified the main culpit as `useSelector` you imported from redux, specifically `stats` and `lastBlogPost`. You can test this by comment out the parts of the page that relied on those variables:

https://github.com/Swetrix/swetrix-fe/blob/main/app/pages/MainPage/index.tsx#L155

Why does this happen? You're calling an endpoint clientside with data the server does not have access to. The server renders the page one way, the client initial render comes out different because of the additional data - bam, hydration error. There are a couple approaches to solving this:

* Wrap the relevant sections inside `<ClientOnly>{() => <>...</></ClientOnly>` from `remix-utils`. This provides a boundary to Remix SSR that says, "ignore everything that happens in here, we'll render it on the client only".
https://github.com/sergiodxa/remix-utils#clientonly
* Better yet, load the data server-side using Remix loaders to avoid hydration errors alltogether: https://remix.run/docs/en/main/tutorials/blog#loading-data

(from what I can tell, `theme` seems to be working properly, good job!)

Hopefully, React will provide a better hydration debugging support in the future. Until then, we're stuck debugging Remix the old fashion way.

**Reply by Blaumaus** [reply]
@xHomu thank you so much for spending your time on debugging my application!
Using Remix loaders / actions is a good suggestion and I'll work on that as well, we're still in process of migration from React and I'm still figuring out how to properly make Redux grab data from loaders (or maybe find another alternative).

I know it's a lot to ask, but maybe if you have some time, can you please take a look at that PR? https://github.com/Swetrix/swetrix-fe/pull/318
I've fixes several sections you pointed out there but still getting the hydration errors, but I'm still getting this link related error on all the pages I open.

For example, there is a page called `/features` - we don't use any dynamic content or Redux there at all but still getting a hydration error and the only difference between SSR and CSR generated HTML are these `<link>` elements and comments generated by Remix -> https://www.diffchecker.com/BiNzTFTQ/

Is it possible that we've set up Remix in a wrong way?

**Reply by xHomu** [reply]
I don't recall seeing that particular links issue during my testing yesterday. Perhaps it's your browser environment? An extension, perhaps? Does it happen if you incognito or another browser?

**Reply by xHomu** [reply]
Actually, it might be the `clearBrowserExtensionInjectionsBeforeHydration()` function in your entry.client, try disabling that.
